packages/websockets/src/server/websocket.server.response.ts

```diff
@@ -0,0 +1,69 @@
+import * as WebSocket from 'ws';
+import * as O from 'fp-ts/lib/Option';
+import * as T from 'fp-ts/lib/Task';
+import * as TE from 'fp-ts/lib/TaskEither';
+import { defer, forkJoin, Observable, of } from 'rxjs';
+import { catchError, mapTo } from 'rxjs/operators';
+import { constant, constVoid, flow, pipe } from 'fp-ts/lib/function';
+import { ContextProvider, Event, LoggerToken, LoggerTag, LoggerLevel } from '@marblejs/core';
+import { EventTransformer } from '../transformer/websocket.transformer.interface';
+import { WebsocketReadyStateMap } from '../websocket.interface';
+import { sendMessage } from './websocket.server.helper';
+
+type ResponseHandlerContext = {
+  eventTransformer: EventTransformer<any>;
+  ask: ContextProvider;
+};
+
+type ClientResponseHandlerContext = ResponseHandlerContext & { client: WebSocket & { id?: string }}
+type ServerResponseHandlerContext = ResponseHandlerContext & { server: WebSocket.Server }
+
+type ClientResponseHandler = (ctx: ClientResponseHandlerContext) => <T extends Event>(event: T) => Observable<boolean>;
+type ServerResponseHandler = (ctx: ServerResponseHandlerContext) => <T extends Event>(event: T) => Observable<boolean>;
+
+const log = (level: LoggerLevel) => (ask: ContextProvider) => (message: string): T.Task<void> =>
+  pipe(
+    ask(LoggerToken),
+    O.map(logger => logger({ level, message, tag: LoggerTag.WEBSOCKETS, type: 'ServerResponseHandler' })),
+    O.getOrElse(constant(constVoid)),
+    T.fromIO);
+
+const logWarn = log(LoggerLevel.WARN);
+const logError = log(LoggerLevel.ERROR);
+
+/**
+ * Try to emit an event to given WebSocket client
+ */
+export const emitEvent: ClientResponseHandler = ({ client, eventTransformer, ask }) => <T extends Event>(event: T) => {
+  const skipSendingEvent: T.Task<boolean> =
+    pipe(
+      logWarn(ask)(`Trying to send an event for client which is ${WebsocketReadyStateMap[client.readyState]}`),
+      T.map(constant(false)),
+    );
+
+  const tryToSendEvent: T.Task<boolean> =
+    pipe(
+      TE.tryCatch(
+        () => pipe(eventTransformer.encode(event), sendMessage(client)),
+        constant(`An error occured while sending and event to client "${client.id}"`)),
+      TE.fold(
+        flow(logError(ask), T.map(constant(false))),
+        constant(T.of(true)),
+    ));
+
+  const isConnectionOpen = client.readyState === WebSocket.OPEN;
+
+  return pipe(
+    defer(pipe(isConnectionOpen ? tryToSendEvent : skipSendingEvent)),
```

Comment:
Just wondering about one thing. 
If I understand correctly sending the event in `emitEvent` was previously eager, but now is lazy, which means that is the event will not be sent until the returned observable is subscribed.

Might be wrong or overthinking it, but isn't it a bit of a breaking change?
Suppose we have an epic like:
```ts
const effect: WsEffect = (e$, client) => e$.pipe(
  matchEvent('foo'),
  tap(() => client.sendResponse('bar')),
  ignoreElements(),
)
// obviously this should be done with `map`/`mapTo` but nevertheless
```
in this case, the `bar` will not be sent, since the observable returned from `sendResponse` will not be subscribed. 🤔

Reply:
True that, good catch. I wanted to fix this behavior but totally forgot about the potential breaking change. I'll mark in the code that this part has to be fixed in the next major release.

Reply:
But from the other side... the function signature was from the very beginning like this:
```typescript
type ClientResponseHandler = (ctx: ClientResponseHandlerContext) => <T extends Event>(event: T) => Observable<boolean>;
type ServerResponseHandler = (ctx: ServerResponseHandlerContext) => <T extends Event>(event: T) => Observable<boolean>;
```

If someone triggered this function via `tap(() => client.sendResponse('bar')),` then it was implemented incorrectly regarding to the function signature. Observables are not subscribed automatically  in `tap` operator. It worked by a mistake. 🤔

Reply:
I agree the signature is still the same, but unfortunately I would still consider it an api change.

I would err on the side of caution, and assume that mistake was probably abused by users using `client.sendResponse` in an imperative fashion. 🤷

Reply:
Additionally the function triggers async tasks that with previous implementation were not kept in check. In order to fix it I have either pack it in `Observable` or change the function signature to `Promise` as a returned value.

Reply:
Or I can leave with it and trigger the side effect separately

Reply:
It will look like this 
```typescript
  // ...
  pipe(isConnectionOpen ? tryToSendEvent : skipSendingEvent)();

  return of(true);
```

Reply:
Right, it should work as previously that way 🤔 
(btw, isn't that `pipe` redundant?)

Reply:
Yeah, I've just copy and pasted it. So I'll leave it as proposed in the latest comment and mark it as a place for fix within next major release.